README.rst
.. contents::


Documentation
=============


What is this?
-------------

This package defines base classes ``BaseCSVReader`` and
``BaseMultilineCSVReader``.  These can be used to iterate over a csv
file and return its contents as a dictionary.  Normally you should use
the ``BaseCSVReader``.  The ``BaseMultilineCSVReader`` can be used
when you run into problems with csv files that can have newline
characters within a column, which could trip up the standard reader.


Example usage
-------------

You should write an own class that inherits from one of the base
classes.  The ``example.py`` file has an example.  Basically it
will be something like this::

  from collective.csv2dict import BaseCSVReader, to_int, to_string

  class ExampleCSVReader(BaseCSVReader):
      """Example csv reader class.

      We read three columns and skip one.
      """
      skip = [2]  # skip column index 2
      fields = [
          # The format is: (field name, filter method)
          ('id', to_int),
          ('fullname', to_string),
          ('email', to_string),
      ]

You can then use this class to read a csv file.  The ``example.py``
file again has sample code to read the csv file and some options from
the command line.  Simply put, it boils down to this::

    c = reader(open(filename, 'U'))
    # Iterate over the entries and print them.
    for entry in c:
        print entry
    print '%d entries ignored due to errors.' % c.ignored
    print '%d entries read without errors.' % c.success

It would turn this csv (contained in ``example.csv``)::

  1,Maurits van Rees,ignored,maurits@example.org
  2,Arthur Dent,ignored again,dentarthurdent@example.org

into this dictionary::

   {'email': u'maurits@example.org',
    'fullname': u'Maurits van Rees',
    'id': 1}
   {'email': u'dentarthurdent@example.org',
    'fullname': u'Arthur Dent',
    'id': 2}


Notes
-----

- It is recommented to always open a file in universal newline mode.
  This is usually the best way to avoid some potential problems with
  newlines within a single row.

- The base reader tries to guess the encoding of the file in a
  simplistic way and will avoid breaking when no good encoding can be
  found.

- The reader might ignore the first row of the csv file as it may be a
  header.  We do a simple check for this: if none of the columns of
  the first row can be turned into an integer, then it is not a header
  line and it will be treated as data.  If this logic does not work
  for you, then override the ``is_header`` method in your own class,
  simply like this::

    def is_header(self, items):
        return False

  That will make sure the first line is always treated as data.  If
  you want it to always be treated as a header, just do ``return
  True``.

- You can override the ``prepare_iterable`` method if you need to
  do some fixes to some rows or the complete csv file before the
  reader starts to handle it.  The ``BaseMultilineCSVReader`` has an
  example for this.

- By default the excel csv dialect is used (or whatever your Python
  version has as default).  If you want to use a specific dialect, you
  can override the ``dialect`` variable in your reader class.  For
  example, you can use tabs as delimiter like this::

    import csv

    class MyDialect(csv.excel):
        delimiter = '\t'

    csv.register_dialect('mydialect', MyDialect)

    class ExampleCSVReader(BaseCSVReader):
        dialect = 'mydialect'
        fields = [...]


Compatibility
-------------

I have tried this on Python 2.6 and an earlier version on 2.4.  It
will likely work on all 2.x versions from 2.3 onwards.

Tested on Mac OS X so likely also working on any Unix-like system.
Should work on Windows too, though I can imagine problems with newline
characters in some corner cases.


Note for Plone users
--------------------

I usually make packages for use in Plone, but this one can be used
with plain Python.  Nevertheless, a note for Plone users is probably
good.

If you want to use it within your Plone buildout, just add it to the
eggs in your buildout.cfg.  You do not need to load zcml or install
anything.  You just need to write your own class definition, as in the
example above.  Then you probably want to write a browser view that
uses this class to turn some uploaded csv file to a dictionary.  Then
you probably create a content item or a member for each item in this
dictionary or do whatever you want with it.


Authors
-------

- Maurits van Rees (package creation, various improvements and
  generalizations)

- Guido Wesdorp (initial code, written for a client way back in 2007)


CHANGES.rst
Changelog
=========


1.2 (unreleased)
----------------

- Nothing changed yet.


1.1 (2014-04-11)
----------------

- Optionally allow ignoring extra columns.  To use this: initialize
  the reader with ``ignore_extra_columns=True``.
  [maurits]

- Add ``formatting`` method to readers.  It currently returns the
  delimiter, the dialect instance, the encoding and the expected
  number of columns.  You can use this to give a hint in an upload
  form.
  [maurits]


1.0 (2012-06-21)
----------------

- Initial release
  [maurits]


collective/csv2dict/utils.py
"""Base importer of CSV files.
"""

import logging
from pprint import pprint

logger = logging.getLogger('collective.csv2dict')
NEWLINE_MARKER = '--NEWLINE--'


def to_string(v):
    v = v.strip()
    if v == '0':
        return ''
    return v


def to_int(v):
    v = v.strip()
    if not v:
        return None
    return int(v)


def to_bool(v):
    v = v.strip()
    return bool(v)


def to_text(v):
    # For fields with multiple lines that have first been changed into
    # NEWLINE_MARKER.
    v = v.strip()
    if not v:
        return ''
    return v.replace(NEWLINE_MARKER, '\n')


def setup_logging():
    # Set up logging.
    formatter = logging.Formatter('[%(levelname)s] %(message)s')
    handler = logging.StreamHandler()
    handler.setFormatter(formatter)
    logger.addHandler(handler)
    logger.setLevel(logging.DEBUG)


def handle_file(reader, filename, lines_to_print=0, raise_exceptions=False,
                print_key=None):
    """Open the file with the reader.

    Arguments:

    - reader: class to open the file with.

    - filename: filename to open.  We open it in universal newlines
      mode, as that potentially avoids lots of errors with Python 2.4
      and some with Python 2.6.

    - lines_to_print: number of entries to print.  By default we print
      nothing, so you can easily check if all lines can be read
      correctly.

    - raise_exceptions: stop when encountering an error.  This value
      is passed to the reader.

    - print_key: when set, we print the value of this key for each
      item in the resulting dictionary.
    """
    c = reader(open(filename, 'U'), raise_exceptions=raise_exceptions)
    logger.info('Reading %s with %s importer', filename, reader.__name__)

    # Test that all entries can be read.  Output some info when wanted.
    for entry in c:
        if print_key:
            print entry.get(print_key)
        if c.success <= lines_to_print:
            pprint(entry)
    logger.info('%d entries ignored due to errors.', c.ignored)
    logger.info('%d entries read without errors.', c.success)


collective/csv2dict/example.py
"""Example importer of a CSV file.
"""

import sys

from readers import BaseCSVReader
from utils import to_int, to_string
from utils import setup_logging, handle_file


class ExampleCSVReader(BaseCSVReader):
    """Example csv reader class.

    We read three columns and skip one.  It would turn this csv:

      1,Maurits van Rees,ignored,maurits@example.org
      2,Arthur Dent,ignored again,dentarthurdent@example.org

    into this dictionary:

      {'email': u'maurits@example.org',
       'fullname': u'Maurits van Rees',
       'id': 1}
      {'email': u'dentarthurdent@example.org',
       'fullname': u'Arthur Dent',
       'id': 2}

    """
    skip = [2]
    fields = [
        # The format is: (field name, filter method)
        ('id', to_int),
        ('fullname', to_string),
        ('email', to_string),
    ]


def usage():
    text = """Usage: %(prog)s <csv filename to import> [num]

This then uses our importer on the specified file.
The goal is mostly to see if the complete file can be read
without errors.

Optionally, you can specify a number of entries that will be printed.
""" % dict(prog=sys.argv[0])
    print text


if __name__ == '__main__':
    setup_logging()
    # Handle command line arguments.
    if len(sys.argv) < 2:
        usage()
        sys.exit(1)
    if len(sys.argv) > 2:
        try:
            lines_to_print = int(sys.argv[2])
        except:
            usage()
            sys.exit(1)
    else:
        lines_to_print = 0

    filename = sys.argv[1]
    handle_file(ExampleCSVReader, filename, lines_to_print=lines_to_print)


collective/csv2dict/readers.py
"""Base importer of CSV files.
"""

import csv
import logging

from utils import NEWLINE_MARKER

# We try a few encodings while reading the csv file; we try to be a
# bit smart about it though.
ENCODINGS = ('ascii', 'iso-8859-1', 'utf-8', 'cp1252')
logger = logging.getLogger('collective.csv2dict')


class CSVImportError(Exception):
    pass


def to_string(v):
    v = v.strip()
    if v == '0':
        return ''
    return v


def to_int(v):
    v = v.strip()
    if not v:
        return None
    return int(v)


def to_bool(v):
    v = v.strip()
    return bool(v)


def to_text(v):
    # For fields with multiple lines that have first been changed into
    # NEWLINE_MARKER.
    v = v.strip()
    if not v:
        return ''
    return v.replace(NEWLINE_MARKER, '\n')


class BaseCSVReader(object):

    # This is the default encoding we try.
    encoding = 'utf-8'
    # The fields, in order of appearance.  Override at least this in
    # your class.
    fields = [
        # The format is: (field name, filter method)
        # Example: ('year', to_int),
    ]
    # Skip the columns with these indexes.  The first column has index 0.
    skip = []
    # Registered csv dialect to use.  With None we use the standard.
    dialect = None

    def prepare_iterable(self, iterable):
        """Do any preparing work to the iterable.

        Override this in child classes if you need to do something
        special, like handling for csv files where there are line
        breaks within a single row.
        """
        return iterable

    def __init__(self, iterable, raise_exceptions=False,
                 ignore_extra_columns=False):
        """Initialize.
        """
        iterable = self.prepare_iterable(iterable)
        if self.dialect is not None:
            self.reader = csv.reader(iterable, self.dialect)
        else:
            self.reader = csv.reader(iterable)
        self.lineno = 0  # for more useful debug info
        self.ignored = 0  # Number of ignored lines
        self.success = 0  # Successfully imported lines
        self.errors = []  # List of errors encountered.
        # Should we raise exceptions or just continue when
        # encountering some bad data?
        self.raise_exceptions = raise_exceptions
        # Can we ignore unexpected extra columns?
        self.ignore_extra_columns = ignore_extra_columns

    def formatting(self):
        info = {
            'delimiter': self.reader.dialect.delimiter,
            'dialect': self.reader.dialect,
            'encoding': self.encoding,
            'ncolumns': len(self.fields) + len(self.skip),
        }
        return info

    def decode_it(self, value):
        """Decode the string value to unicode.

        We don't like to fail due to some unicode problems, so we try
        to be smart about choosing a nice encoding.
        """
        # First try the standard encoding that is set in the class.
        try:
            value = value.decode(self.encoding)
        except UnicodeDecodeError:
            pass
        else:
            return value
        # When that fails, try the supported encodings.
        for encoding in ENCODINGS:
            try:
                value = value.decode(self.encoding)
            except UnicodeDecodeError:
                pass
            else:
                # Make sure this encoding gets tried first for the next value:
                self.encoding = encoding
                return value
        # When all else fails, just ignore the errors.
        return value.decode('utf-8', 'ignore')

    def is_header(self, items):
        # We do a simple check: if none of the items can be turned into
        # an integer, then it is not a header line.
        for item in items:
            try:
                int(item.strip())
            except:
                continue
            # We have an integer, which will not happen in a header.
            return False
        # No integers found, so it is a header.
        return True

    def next(self):
        """Get the next line.

        It could be a borked line though, with 'newline inside string'
        (Python 2.4) or 'new-line character seen in unquoted field'
        (Python 2.6).  When that happens, Python suggests:

          "do you need to open the file in universal-newline mode?"

        So that is this: open(filename, 'U')

        In case of borked lines, the next few lines could also be
        broken due to the same error or due to not enough columns, so
        we try a few times to get a working line.
        """

        expected_columns = len(self.fields) + len(self.skip)
        tries = 0
        max_tries = 100
        while tries < max_tries:
            tries += 1
            self.lineno += 1
            try:
                items = self.reader.next()
            except StopIteration:
                raise
            except Exception, e:
                self.ignored += 1
                msg = "%s: line %d (row %d) ignored because of error: %s" % (
                    self.__class__.__name__, self.lineno,
                    self.ignored + self.success, e)
                self.errors.append(msg)
                logger.warn(msg)
                if self.raise_exceptions:
                    raise CSVImportError(msg)
                continue
            else:
                if len(items) == expected_columns:
                    # A good line found
                    break
                elif self.ignore_extra_columns \
                        and len(items) > expected_columns:
                    # A good enough line found.
                    if self.lineno == 1:
                        logger.warn(
                            "No correct number of columns (%d instead of %d)."
                            % (len(items), expected_columns))
                    else:
                        # Do the extra columns contain data?
                        data = ''.join(items[expected_columns:])
                        if data:
                            extra = self.reader.dialect.delimiter.join(
                                items[expected_columns:])
                            logger.warn(
                                "Line %d has incorrect number of columns "
                                "(%d instead of %d) and they contain data: %r"
                                % (self.lineno, len(items), expected_columns,
                                   extra))
                    break
                elif tries == 1:
                    # Bad line found on first try.
                    self.ignored += 1
                    error = ("No correct number of columns (%d instead of %d)."
                             % (len(items), expected_columns))
                    msg = ("%s: line %d (row %d) ignored because of error: "
                           "%s" % (
                               self.__class__.__name__, self.lineno,
                               self.ignored + self.success, error))
                    logger.warn(msg)
                    if self.raise_exceptions:
                        raise CSVImportError(msg)
                else:
                    # Bad line found, but caused by an earlier reading
                    # error, likely a 'newline inside string'
                    # exception, so we have already counted this as an
                    # ignored row.
                    pass
        if tries >= max_tries:
            raise CSVImportError("Failed to find a good line after %d tries"
                                 % tries)
        # First line might be a header, we ignore that one then.
        if self.lineno == 1 and self.is_header(items):
            self.lineno += 1
            items = self.reader.next()
        ret = {}
        if self.skip:
            # Some columns need to be skipped.
            items = [item for (index, item) in enumerate(items)
                     if index not in self.skip]
        for index, (name, filter) in enumerate(self.fields):
            value = items[index]
            if value == '\N':
                # This is NULL.
                value = None
            if filter and value is not None:
                try:
                    value = filter(value)
                except Exception, e:
                    # Make the exception info a little more useful or
                    # ignore it.
                    msg = '%s (field="%s", value="%s", line %d, filter %r)' % (
                        e, name, value, self.lineno, filter.__name__)
                    self.errors.append(msg)
                    logger.warn(msg)
                    if self.raise_exceptions:
                        raise CSVImportError(msg)
                    value = None
            if isinstance(value, str):
                value = self.decode_it(value)
            ret[name] = value
        self.success += 1
        return ret

    def __iter__(self):
        return self


class BaseMultilineCSVReader(BaseCSVReader):
    """Read rows that span multiple lines.

    Use this when you run into problems with the base reader.  Note
    that Python 2.6 is already better than 2.4 here, and opening files
    in universal newline mode also helps.
    """

    def prepare_iterable(self, iterable):
        """Do any preparing work to the iterable.

        Sometimes there is just too much going wrong with newlines, so
        we need to get in and fix it ourselves.  Quite ugly.

        We could make this a generator and use yield, but it is a bit
        hard to know when exactly a line is correct.
        """
        new = []
        broken = False
        for line in iterable.readlines():
            if line == '\\\n':
                # The next line belongs to the previous line.
                broken = True
                continue
            line = line.strip()
            if broken:
                # This line belongs to the previous line.
                new[-1] += NEWLINE_MARKER + line
                broken = False
            else:
                # All is normal.
                new.append(line)
            if line.endswith('\\'):
                # There will be a continuation in the next line.
                broken = True
                continue
        return new
